src/Helpers/Arr.php

```diff
@@ -71,4 +71,26 @@ public static function get(array $array, string|int|null $key, mixed $default =
 
         return $array;
     }
+
+    /**
+     * Map an array with keys
+     *
+     * @param array $items
+     * @param callable $callback
+     * @return array
+     */
+    public static function mapWithKeys(array $items, callable $callback): array
+    {
+        $result = [];
+
+        foreach ($items as $key => $value) {
+            $assoc = $callback($value, $key);
+
+            foreach ($assoc as $mapKey => $mapValue) {
+                $result[$mapKey] = $mapValue;
+            }
+        }
+
+        return $result;
+    }
```

Comment:
Another small suggestion so you don't have to nest loops -
I think you can 'dumb' this down a bit, as we're expecting the `$callback` to return an array.

```suggestion
    public static function mapWithKeys(array $items, callable $callback): array
    {
        $result = [];

        foreach ($items as $key => $value) {
            $result += $callback($value, $key);
        }

        return $result;
    }
```

A side note, though  - the above will keep the first value, whereas the current behaviour will overwrite - so maybe `array_merge()` instead of union is better.

Reply:
I think I'll leave this one as this was copied from Laravel's code, and pretty happy with how it works